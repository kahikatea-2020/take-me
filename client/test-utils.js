import React from 'react'
import { Provider } from 'react-redux'
import { createStore, applyMiddleware, compose } from 'redux'
import { MemoryRouter as Router } from 'react-router-dom'
import thunkMiddleware from 'redux-thunk'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import reducer from './reducers'
const enhancer = compose(applyMiddleware(thunkMiddleware))

export const renderWithRedux = (
  ui,
  { initialState, initialEntries = ['/'], store = createStore(reducer, initialState, enhancer) } = {}
) => {
  return {
    ...render(
      <Provider store={store}>
        <Router initialEntries={initialEntries} initialIndex={0}>
          {ui}
        </Router>
      </Provider>
    ),
    store
  }
}
