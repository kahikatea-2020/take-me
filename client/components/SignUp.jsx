import React from 'react'
import { Link } from 'react-router-dom'
import { Form } from 'semantic-ui-react'
import { isAuthenticated, register } from 'authenticare/client'
import { connect } from 'react-redux'

import { openUploadWidget } from './CloudinaryService'
import { showError, hideError } from '../actions/error'
import { BASE_API_URL } from '../base-api.js'
import { userPending, userSuccess, getUserDetails } from '../actions/users'
import WaitIndicator from './WaitIndicator'

class SignUp extends React.Component {
  state = {
    firstName: '',
    lastName: '',
    emailAddress: '',
    phoneNumber: null,
    location: '',
    username: '',
    password: '',
    confirmPassword: '',
    imageUrl: ''
  }

  inputChecker = event => {
    const { firstName, lastName, emailAddress, location, username, password, phoneNumber } = this.state
    if (firstName !== '' && lastName !== '' && location !== '' && username !== '' && password !== '' && emailAddress !== '') {
      if (phoneNumber !== null || phoneNumber !== '') {
        return false
      } else {
        return false
      }
    } else {
      return true
    }
  }

  imageUpload = tag => {
    const uploadOptions = {
      cloudName: 'takemenz',
      tags: [tag],
      uploadPreset: 'nxxqgset'
    }

    openUploadWidget(uploadOptions, (error, photo) => {
      if (!error) {
        if (photo.event === 'success') {
          this.setState({
            imageUrl: photo.info.path
          })
        }
      }
    })
  }

  updateField = e => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  submitHandler = e => {
    this.props.dispatch(hideError())
    if (this.state.password !== this.state.confirmPassword) {
      this.props.dispatch(showError('Password does not match'))
    } else if (this.inputChecker()) {
      this.props.dispatch(showError('Please fill out all the fields'))
    } else {
      this.props.dispatch(userPending())
      register(this.state, { baseUrl: BASE_API_URL })
        .then((token) => {
          this.props.dispatch(userSuccess())
          if (isAuthenticated()) {
            this.props.dispatch(getUserDetails())
          }
        })
        .catch(() => {
          this.props.dispatch(userSuccess())
          this.props.dispatch(showError('Username already taken'))
        })
        .then(() => {
          this.props.history.push('/')
        })
    }
  }

  render () {
    return (
      <>
        <h1>Sign Up</h1>
        <p>Please fill in the following details:</p>
        {this.props.error && <div>{this.props.error}</div>}
        <Form>
          <Form.Input
            onKeyUp={this.updateField}
            fluid
            required
            width={6}
            name='firstName'
            placeholder='First name'
            type='text'
          />
          <Form.Input
            onKeyUp={this.updateField}
            fluid
            required
            width={6}
            name='lastName'
            placeholder='Last name'
            type='text'
          />
          <Form.Input
            onKeyUp={this.updateField}
            fluid
            required
            width={6}
            name='emailAddress'
            placeholder='Email address'
            type='text'
          />
          <Form.Input
            onKeyUp={this.updateField}
            fluid
            required
            width={6}
            name='phoneNumber'
            placeholder='Phone number'
            type='number'
          />
          <Form.Input
            onKeyUp={this.updateField}
            fluid
            required
            width={6}
            name='location'
            placeholder='Location'
            type='text'
          />
          <Form.Input
            onKeyUp={this.updateField}
            fluid
            required
            width={6}
            name='username'
            placeholder='Username'
            type='text'
          />
          <Form.Input
            onKeyUp={this.updateField}
            fluid
            required
            width={6}
            name='password'
            placeholder='Password'
            type='password'
            autoComplete='off'
          />
          <Form.Input
            onKeyUp={this.updateField}
            fluid
            required
            width={6}
            name='confirmPassword'
            placeholder='Confirm password'
            type='password'
            autoComplete='off'
          />
          <Form.Button onClick={() => this.imageUpload()}>Upload Image</Form.Button>
          {(this.state.imageUrl !== '') &&
            <div className='imagesPreview'>
              <div><img src={`https://res.cloudinary.com/takemenz/image/upload/${this.state.imageUrl}`}/></div>
            </div>
          }
          <Form.Group>
            <Link to='/'>
              <Form.Button>
              Cancel
              </Form.Button>
            </Link>
            <Form.Button
              type='submit'
              onClick={this.submitHandler}
            >
            Submit
            </Form.Button>
          </Form.Group>
        </Form>
        <WaitIndicator />
      </>
    )
  }
}

const mapStateToProps = state => {
  return {
    error: state.error
  }
}

export default connect(mapStateToProps)(SignUp)
