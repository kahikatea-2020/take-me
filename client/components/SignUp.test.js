import React from 'react'
import { shallow } from 'enzyme'
import { Form } from 'semantic-ui-react'

import { renderWithRedux } from '../test-utils'
import '@testing-library/jest-dom'

import SignUp from './SignUp'

test('<SignUp /> renders correctly', () => {
  const rendered = renderWithRedux(<SignUp />)
  expect(rendered).toMatchSnapshot()
})

test('<SignUp /> has the correct number of inputs', () => {
  const wrapper = shallow(<SignUp />)
  expect(wrapper.find(Form.Input).length).toBe(8)
})
