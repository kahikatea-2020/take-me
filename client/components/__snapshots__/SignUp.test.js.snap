// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="wrapper"
      >
        <h1>
          Sign Up
        </h1>
        <p>
          Please fill in the following details:
        </p>
        <form
          class="ui form"
        >
          <div
            class="required six wide field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="firstName"
                placeholder="First name"
                required=""
                type="text"
              />
            </div>
          </div>
          <div
            class="required six wide field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="lastName"
                placeholder="Last name"
                required=""
                type="text"
              />
            </div>
          </div>
          <div
            class="required six wide field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="emailAddress"
                placeholder="Email address"
                required=""
                type="text"
              />
            </div>
          </div>
          <div
            class="required six wide field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="phoneNumber"
                placeholder="Phone number"
                required=""
                type="number"
              />
            </div>
          </div>
          <div
            class="required six wide field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="location"
                placeholder="Location"
                required=""
                type="text"
              />
            </div>
          </div>
          <div
            class="required six wide field"
          >
            <div
              class="ui fluid input"
            >
              <input
                name="username"
                placeholder="Username"
                required=""
                type="text"
              />
            </div>
          </div>
          <div
            class="required six wide field"
          >
            <div
              class="ui fluid input"
            >
              <input
                autocomplete="off"
                name="password"
                placeholder="Password"
                required=""
                type="password"
              />
            </div>
          </div>
          <div
            class="required six wide field"
          >
            <div
              class="ui fluid input"
            >
              <input
                autocomplete="off"
                name="confirmPassword"
                placeholder="Confirm password"
                required=""
                type="password"
              />
            </div>
          </div>
          <div
            class="field"
          >
            <button
              class="ui button"
            >
              Upload Image
            </button>
          </div>
          <div
            class="fields"
          >
            <a
              href="/"
            >
              <div
                class="field"
              >
                <button
                  class="ui button"
                >
                  Cancel
                </button>
              </div>
            </a>
            <div
              class="field"
            >
              <button
                class="ui button"
                type="submit"
              >
                Submit
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="wrapper"
    >
      <h1>
        Sign Up
      </h1>
      <p>
        Please fill in the following details:
      </p>
      <form
        class="ui form"
      >
        <div
          class="required six wide field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="firstName"
              placeholder="First name"
              required=""
              type="text"
            />
          </div>
        </div>
        <div
          class="required six wide field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="lastName"
              placeholder="Last name"
              required=""
              type="text"
            />
          </div>
        </div>
        <div
          class="required six wide field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="emailAddress"
              placeholder="Email address"
              required=""
              type="text"
            />
          </div>
        </div>
        <div
          class="required six wide field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="phoneNumber"
              placeholder="Phone number"
              required=""
              type="number"
            />
          </div>
        </div>
        <div
          class="required six wide field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="location"
              placeholder="Location"
              required=""
              type="text"
            />
          </div>
        </div>
        <div
          class="required six wide field"
        >
          <div
            class="ui fluid input"
          >
            <input
              name="username"
              placeholder="Username"
              required=""
              type="text"
            />
          </div>
        </div>
        <div
          class="required six wide field"
        >
          <div
            class="ui fluid input"
          >
            <input
              autocomplete="off"
              name="password"
              placeholder="Password"
              required=""
              type="password"
            />
          </div>
        </div>
        <div
          class="required six wide field"
        >
          <div
            class="ui fluid input"
          >
            <input
              autocomplete="off"
              name="confirmPassword"
              placeholder="Confirm password"
              required=""
              type="password"
            />
          </div>
        </div>
        <div
          class="field"
        >
          <button
            class="ui button"
          >
            Upload Image
          </button>
        </div>
        <div
          class="fields"
        >
          <a
            href="/"
          >
            <div
              class="field"
            >
              <button
                class="ui button"
              >
                Cancel
              </button>
            </div>
          </a>
          <div
            class="field"
          >
            <button
              class="ui button"
              type="submit"
            >
              Submit
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
  "unmount": [Function],
}
`;
